### PR TITLE
Add nested support for dense_vector fields and knn search

### PR DESCRIPTION
**Pull request by benwtrent**
Adds `nested` support for `dense_vector`. 

The stipulations are:
 - `inner_hits` only ever returns one hit, the nearest vector document
 - `filter` within kNN is a filter over parent docs only
 - Results are diversified over parent docs. So you get `k` documents scored by their nearest passage

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 1]
```diff
@@ -139,8 +140,10 @@ public static List<DfsKnnResults> mergeKnnResults(SearchRequest request, List<Df
         }
 
         List<List<TopDocs>> topDocsLists = new ArrayList<>(request.source().knnSearch().size());
+        List<SetOnce<String>> nestedPath = new ArrayList<>(request.source().knnSearch().size());
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:143]: What's the reason we need SetOnce here, why not just List<String> ?

**Review by mayya-sharipova — APPROVED** [review]
@benwtrent Thanks Ben for a great work, this new functionality would be very much welcome!

I still have some concerns though:
-  It was surprising for me that for  nested knn search we did not use "nested" keyword, the way we used "nested" queries for "nested" field types (I understand that `knn` is a top level section). 
- If we decide to go with knn as a query,  what would be a difference of this nested `knn` search VS  a `nested` query with `knn` query ?

**Comment by benwtrent** [comment]
> It was surprising for me that for nested knn search we did not use "nested" keyword, the way we used "nested" queries for "nested" field types (I understand that knn is a top level section).

"nested" where @mayya-sharipova ? I could see us adding a `nested_filter` or something that allows filtering by other nested metadata. But using `filter` (which is a pre-filter) that filters over parent level docs made the most sense and would be the most common use-case.

> If we decide to go with knn as a query, what would be a difference of this nested knn search VS a nested query with knn query ?

I would expect `inner_hits` within the `nested` query is where those would be defined, not within the `knn` query directly. 

Additionally, we may require a `nested` query in the query DSL when using `knn` search against a nested `dense_vector` field (just like we do all other nested fields).

As for how the prefiltering works for a dsl `knn` query, I could easily see it being used in the same way here. But we can discuss if that makes the most sense or not.

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java (hunk 1 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:143]: Honestly, I am thinking we should have an assertion with the `setonce`. We should ONLY see a single nested path (as we are querying the same knn field over all the indices/shards). Seeing more than one nested path for the same knn clause declaration would be bad.

**Comment by benwtrent** [comment]
> I'm wondering if there's any plan to support inner_hits > 1 in the roadmap or if there are technical limitations hindering its implementation? Such a capability would prove valuable in scenarios where multiple passages of a parent document contain relevant information for a user query.

@dbuades could you provide an example usecase? It is definitely technically feasible. I am wondering around what in particular would be best. Inner hits may be the way to do it, but we should also feel free to think of something thats better :D.

**Comment by benwtrent** [comment]
Nice idea @dbuades ! Could you open an GH issue and tag me in it?

**Comment by dbuades** [comment]
Thank you @benwtrent! Sorry that I missed your reply on Friday, I've now created [the issue here](https://github.com/elastic/elasticsearch/issues/102950).

**Pull request merged**